File: src/platform/errors/PlatformError.ts
export type PlatformErrorCode =
    | 'CANCELLED'
    | 'PERMISSION_DENIED'
    | 'INVALID_PATH'
    | 'NOT_FOUND'
    | 'COMMAND_UNAVAILABLE'
    | 'IO_ERROR'
    | 'FILE_TOO_LARGE'
    | 'WATCHER_ERROR'
    | 'INTERNAL_ERROR';

export interface PlatformErrorPayload {
    code: PlatformErrorCode;
    message: string;
}

const PLATFORM_ERROR_CODES = new Set<PlatformErrorCode>([
    'CANCELLED',
    'PERMISSION_DENIED',
    'INVALID_PATH',
    'NOT_FOUND',
    'COMMAND_UNAVAILABLE',
    'IO_ERROR',
    'FILE_TOO_LARGE',
    'WATCHER_ERROR',
    'INTERNAL_ERROR'
]);

export class PlatformError extends Error {
    readonly code: PlatformErrorCode;

    constructor(code: PlatformErrorCode, message: string) {
        super(message);
        this.name = 'PlatformError';
        this.code = code;
    }

    static fromUnknown(error: unknown): PlatformError {
        if (error instanceof PlatformError) {
            return error;
        }
        if (isPlatformErrorPayload(error)) {
            return new PlatformError(error.code, error.message);
        }
        if (error instanceof Error) {
            const unavailable = error.message.includes('not found') || error.message.includes('unknown command');
            return new PlatformError(unavailable ? 'COMMAND_UNAVAILABLE' : 'INTERNAL_ERROR', error.message);
        }
        return new PlatformError('INTERNAL_ERROR', String(error));
    }
}

function isPlatformErrorPayload(value: unknown): value is PlatformErrorPayload {
    if (!value || typeof value !== 'object') {
        return false;
    }
    const candidate = value as Partial<PlatformErrorPayload>;
    return typeof candidate.message === 'string'
        && typeof candidate.code === 'string'
        && PLATFORM_ERROR_CODES.has(candidate.code as PlatformErrorCode);
}

File: src-tauri/src/platform_error.rs
use serde::Serialize;
use std::fmt::{Display, Formatter};

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlatformError {
    pub code: &'static str,
    pub message: String,
}

impl PlatformError {
    pub fn invalid_path(message: impl Into<String>) -> Self {
        Self::new("INVALID_PATH", message)
    }

    pub fn permission_denied(message: impl Into<String>) -> Self {
        Self::new("PERMISSION_DENIED", message)
    }

    pub fn not_found(message: impl Into<String>) -> Self {
        Self::new("NOT_FOUND", message)
    }

    pub fn io(message: impl Into<String>) -> Self {
        Self::new("IO_ERROR", message)
    }

    pub fn too_large(message: impl Into<String>) -> Self {
        Self::new("FILE_TOO_LARGE", message)
    }

    pub fn watcher(message: impl Into<String>) -> Self {
        Self::new("WATCHER_ERROR", message)
    }

    pub fn internal(message: impl Into<String>) -> Self {
        Self::new("INTERNAL_ERROR", message)
    }

    fn new(code: &'static str, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
        }
    }
}

impl Display for PlatformError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}: {}", self.code, self.message)
    }
}

impl std::error::Error for PlatformError {}
